src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+
+      if (fread(framed_entries.data(), framed_size, 1, file) != 1)
+      {
+        throw std::logic_error(fmt::format(
+          "Failed to read entry range {} - {} from file", from, to));
+      }
+
+      return framed_entries;
+    }
+
+    bool truncate(size_t idx)
+    {
+      if (compacted || (idx < start_idx - 1) || (idx >= get_last_idx()))
+      {
+        return false;
+      }
+
+      if (idx == start_idx - 1)
+      {
+        // Truncating everything triggers file deletion
+        if (!fs::remove(fs::path(dir) / fs::path(get_file_name())))
+        {
+          throw std::logic_error(
+            fmt::format("Could not remove file {}", get_file_name()));
+        }
+        return true;
+      }
+
+      // Reset positions offset header
+      fseeko(file, 0, SEEK_SET);
+      positions_offset_header_t table_offset = 0;
+      if (fwrite(&table_offset, sizeof(table_offset), 1, file) != 1)
+      {
+        throw std::logic_error("Failed to reset positions table offset");
+      }
+
+      completed = false;
+      total_len = positions.at(idx - start_idx + 1);
+      positions.resize(idx - start_idx + 1);
+
+      if (fflush(file) != 0)
+      {
+        throw std::logic_error(fmt::format("Failed to flush ledger file"));
+      }
+
+      if (ftruncate(fileno(file), total_len))
+      {
+        throw std::logic_error("Failed to truncate ledger");
+      }
+
+      fseeko(file, total_len, SEEK_SET);
+      return false;
+    }
+
+    void complete()
+    {
+      if (completed)
+      {
+        return;
+      }
+
+      fseeko(file, total_len, SEEK_SET);
+      size_t table_offset = ftello(file);
+
+      if (
+        fwrite(
+          reinterpret_cast<uint8_t*>(positions.data()),
+          sizeof(positions.at(0)),
+          positions.size(),
+          file) != positions.size())
+      {
+        throw std::logic_error("Failed to write positions table to ledger");
+      }
+
+      // Write positions table offset at start of file
+      if (fseeko(file, 0, SEEK_SET) != 0)
+      {
+        throw std::logic_error("Failed to set file offset to 0");
+      }
+
+      if (fwrite(&table_offset, sizeof(table_offset), 1, file) != 1)
+      {
+        throw std::logic_error("Failed to write positions table to ledger");
+      }
+
+      if (fflush(file) != 0)
+      {
+        throw std::logic_error(fmt::format("Failed to flush ledger file"));
+      }
+
+      completed = true;
+    }
+
+    bool compact(size_t idx)
+    {
+      if (!completed || compacted || (idx != get_last_idx()))
+      {
+        // No effect if compaction idx is not last idx
+        return false;
+      }
+
+      if (fflush(file) != 0)
+      {
+        throw std::logic_error(fmt::format("Failed to flush ledger file"));
+      }
+
+      fs::rename(
+        fs::path(dir) / fs::path(get_file_name()),
+        fs::path(dir) /
+          fs::path(fmt::format(
+            "{}-{}{}",
+            get_file_name(),
```

Comment:
Just to confirm - `get_file_name()` includes the start-index, right? So this produces something like `ledger_100-200.committed`?

Reply:
Yes, that's the idea. The start index isn't useful when reading, only the end index is used, but it's nice to have for an operator wanting to eyeball some files.

Reply:
Yup. It's actually the `start_idx` that we care about when we're reading files from disk, the `last_idx` is there for the reason @achamayou  mentioned.